### DISCUSSION TITLE
Unittest indefinitely hangs when using BaseHTTPMiddleware, sync router and quit(). Needs MainThread _daemonic hack.

### DISCUSSION THREAD

**Discussion by arrmansa** [fastapi/fastapi#11254] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import unittest
from fastapi import FastAPI
from fastapi.applications import BaseHTTPMiddleware
from fastapi.testclient import TestClient

class NothingMiddleware(BaseHTTPMiddleware):
    def __init__(self,app,):
        super().__init__(app)
    async def dispatch(self, request, call_next):
        return await call_next(request)

app = FastAPI()
app.add_middleware(NothingMiddleware)
@app.get("/quit")
def just_quit():
    quit(1)

class TestQuitEndpoint(unittest.TestCase):

    def test_SystemExit(self):
        # Crash fix for allowing assertRaises SystemExit - Uncomment next 2 lines for testcase process to exit
        # import threading 
        # setattr(threading.main_thread(), "_daemonic", True)
        # Testcase
        with self.assertRaises(RuntimeError):
            with TestClient(app) as client:
                with self.assertRaises(SystemExit):
                    client.get("/quit")

if __name__ == '__main__':
    unittest.main()
```

### Description

When running the testcase code - if I use `async def just_quit`, it works as expected, but when I do def `just_quit()` I have to use the hack to get the testcase to exit  or it hangs indefinitely.
```python
    import threading
    setattr(threading.main_thread(), "_daemonic", True)
```
`uvicorn app:app works` works and quits as expected either ways. 

### Operating System

macOS

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

Python 3.11.5

### Additional Context

![untitled-2](https://github.com/tiangolo/fastapi/assets/41120982/7b62f7f6-55c6-4623-ada5-968e2955a515)
This is what happens in the hang when I use ctrl+c to exit

**Comment by arrman-saha** [comment]
I originally thought this was a fastapi issue, but seems like this is a [starlette issue](https://github.com/encode/starlette/discussions/2537).
